src/PubSubClient.cpp

```diff
@@ -496,21 +496,21 @@ bool PubSubClient::publish_P(const char* topic, const uint8_t* payload, size_t p
 }
 
 bool PubSubClient::beginPublish(const char* topic, size_t plength, bool retained) {
-    if (connected()) {
+    // check if the header and the topic (including 2 length bytes) fit into the buffer
+    if (connected() && MQTT_MAX_HEADER_SIZE + strlen(topic) + 2 <= this->bufferSize) {
         // Send the header and variable length field
-        size_t length = MQTT_MAX_HEADER_SIZE;
-        length = writeString(topic, this->buffer, length);
+        size_t topicLen = writeString(topic, this->buffer, MQTT_MAX_HEADER_SIZE) - MQTT_MAX_HEADER_SIZE;
```

Comment:
Maybe good idea to write down what is being done here.

The buffer is considered to be split in several sections:
- [0 ... MQTT_MAX_HEADER_SIZE>  Header is "right aligned", thus start at (MQTT_MAX_HEADER_SIZE - <returned size>) bytes into the buffer.
- [MQTT_MAX_HEADER_SIZE ... MQTT_MAX_HEADER_SIZE +1]  2 bytes topic length
- Topic, start at pos (MQTT_MAX_HEADER_SIZE + 2)

Since the length of the topic and payload needs to be announced in the header, we copy the topic first and then pre-pend the header.
The header is of variable length due to the total length being stored as multi-byte value. (7-bit per byte + MSB to signal extra byte)

Reply:
Thanks. I added some more comments here. Hope this helps.